Sources/Algorithms/MinMax.swift

```diff
@@ -383,3 +383,117 @@ extension Collection where Element: Comparable {
+  /// hold:
+  ///
+  /// - `areInIncreasingOrder(a, a)` is always `false`. (Irreflexivity)
+  /// - If `areInIncreasingOrder(a, b)` and `areInIncreasingOrder(b, c)` are
+  ///   both `true`, then `areInIncreasingOrder(a, c)` is also
+  ///   `true`. (Transitive comparability)
+  /// - Two elements are *incomparable* if neither is ordered before the other
+  ///   according to the predicate. If `a` and `b` are incomparable, and `b`
+  ///   and `c` are incomparable, then `a` and `c` are also incomparable.
+  ///   (Transitive incomparability)
+  ///
+  /// This example shows how to use the `minAndMax(by:)` method on a
+  /// dictionary to find the key-value pair with the lowest value and the pair
+  /// with the highest value.
+  ///
+  ///     let hues = ["Heliotrope": 296, "Coral": 16, "Aquamarine": 156]
+  ///     if let extremeHues = hues.minAndMax(by: {$0.value < $1.value}) {
+  ///         print(extremeHues.min, extremeHues.max)
+  ///     } else {
+  ///         print("There are no hues")
+  ///     }
+  ///     // Prints: "(key: "Coral", value: 16) (key: "Heliotrope", value: 296)"
+  ///
+  /// - Precondition: The sequence is finite.
+  ///
+  /// - Parameter areInIncreasingOrder: A predicate that returns `true`
+  ///   if its first argument should be ordered before its second
+  ///   argument; otherwise, `false`.
+  /// - Returns: A tuple with the sequence's minimum element, followed by its
+  ///   maximum element. For either member, if the sequence provides multiple
+  ///   qualifying elements, the one chosen is unspecified. The same element may
+  ///   be used for both members if all the elements are equivalent. If the
+  ///   sequence has no elements, returns `nil`.
+  ///
+  /// - Complexity: O(*n*), where *n* is the length of the sequence.
+  public func minAndMax(
+    by areInIncreasingOrder: (Element, Element) throws -> Bool
+  ) rethrows -> (min: Element, max: Element)? {
+    var iterator = makeIterator()
+    guard var lowest = iterator.next() else { return nil }
+
+    var highest = iterator.next() ?? lowest
+    if try areInIncreasingOrder(highest, lowest) {
+      swap(&lowest, &highest)
+    }
+    while let element = iterator.next() {
+      guard let following = iterator.next() else {
+        if try areInIncreasingOrder(element, lowest) {
+          lowest = element
+        } else if try !areInIncreasingOrder(element, highest) {
+          highest = element
+        }
+        break
+      }
+
+      if try areInIncreasingOrder(following, element) {
+        if try areInIncreasingOrder(following, lowest) {
+          lowest = following
+        }
+        if try !areInIncreasingOrder(element, highest) {
+          highest = element
+        }
+      } else {
+        if try areInIncreasingOrder(element, lowest) {
+          lowest = element
+        }
+        if try !areInIncreasingOrder(following, highest) {
+          highest = following
+        }
+      }
+    }
+    return (lowest, highest)
+  }
```

Comment:
The while loop can be substantially simplified, since iterators are guaranteed to keep returning `nil` after they’ve done so once:

```swift
while var low = iterator.next() {
  var high = iterator.next() ?? low
  if try areInIncreasingOrder(high, low) { swap(&low, &high) }
  if try areInIncreasingOrder(low, lowest) { lowest = low }
  if try !areInIncreasingOrder(high, highest) { highest = high }
}
```

This does technically perform one extraneous comparison on odd-length sequences (`high` vs `low` after the `??` was triggered), but even that can be rectified by restoring the `guard` if benchmarks show a measurable difference:

```swift
while var low = iterator.next() {
  guard var high = iterator.next() else {
    if try areInIncreasingOrder(low, lowest) { lowest = low }
    if try !areInIncreasingOrder(low, highest) { highest = low }
    break
  }
  if try areInIncreasingOrder(high, low) { swap(&low, &high) }
  if try areInIncreasingOrder(low, lowest) { lowest = low }
  if try !areInIncreasingOrder(high, highest) { highest = high }
}
```